### ISSUE TITLE
E1101, E1136 false positive on dataclasses fields with default_factory (no-member, unsubscriptable-object)

### ISSUE THREAD

**Issue by sedrubal** [PyCQA/pylint#2605]
### Steps to reproduce
1. Use this script (test.py):
```py
"""A test script the confuses pylint."""
from dataclasses import dataclass, field


@dataclass
class Test:
    """A test dataclass with a field, that has a default_factory."""

    test: list = field(default_factory=list)


TEST = Test()
TEST.test.append(1)
print(TEST.test[0])
```
2. run `pylint test.py`

### Current behavior
pylint complains about the member `test`, because dataclasses change the type of it during runtime. But there is no other way when you want to use datalcasses:

```
************* Module test
test.py:13:0: E1101: Instance of 'Field' has no 'append' member (no-member)
test.py:14:6: E1136: Value 'TEST.test' is unsubscriptable (unsubscriptable-object)

--------------------------------------------------------------------
Your code has been rated at -6.67/10 (previous run: -6.67/10, +0.00)
```

### Expected behavior
There shouldn't be any error or warning in this example script.

### pylint --version output
```
pylint-3 2.1.1
astroid 2.0.4
Python 3.7.1 (default, Nov  5 2018, 14:07:04)
[GCC 8.2.1 20181011 (Red Hat 8.2.1-4)]
```

**Comment by PCManticore** [comment]
Thanks for reporting an issue! Currently pylint lacks support for understanding `dataclasses` fully, one such area where it currently lacks being dataclasses factories.

**Comment by OrBin** [comment]
I have the same issue for a member of type `dict` with value `field(default_factory=list)`.

Class definition:
```python
@dataclass
class TriggerResult:
    should_respond: bool
    response_payload: dict = field(default_factory=dict)
```

Usage (here, inside a pytest test case):
```python
# result is a variable of type TriggerResult
assert 'match' in result.response_payload
match = result.response_payload['match']
```

I get the following errors:
* E1135 (unsupported-membership-test) for an `in` operator in the second line of the snippet
* E1136 (unsubscriptable-object) for a `[]` operator in the third line of the snippet

**Comment by MarcBoissonneault** [comment]
This issue is the same as this one (both are related to the type used in the `defautl_factory` fo `dataclass.field`): https://github.com/PyCQA/pylint/issues/2698

**Comment by AWhetter** [comment]
Looks like this might have been fixed by https://github.com/PyCQA/astroid/commit/3e4eac76e9abfa0400fa07b58c61973ac5495bd2 ?
I cannot reproduce.

**Comment by OrBin** [comment]
> Looks like this might have been fixed by [PyCQA/astroid@3e4eac7](https://github.com/PyCQA/astroid/commit/3e4eac76e9abfa0400fa07b58c61973ac5495bd2) (astroid v2.3.1)?
> I cannot reproduce.

Seems fixed to me. I no longer get those errors I reported before (E1135 unsupported-membership-test and E1136 unsubscriptable-object).
I believe this issue should be closed now.

**Comment by AWhetter** [comment]
Thanks for the feedback. Closing!

**Comment by tsvisab** [comment]
Reproduced with `Astroid 2.5.2`, `pylint  2.7.4`
A nasty workaround if you must is to explicitly cast the dataclass field into the subscriptable type:

`
test_list = list(TEST.test)
test_list.append(1)
print(test_list.test[0])
`

**Comment by Pierre-Sassoulas** [comment]
This will be fixed in the next release maybe already fixed with astroid 2.5.3.
